packages/material-ui/src/internal/SwitchBase.js

```diff
@@ -30,128 +29,117 @@ export const styles = {
-    }
-  }
-
-  handleFocus = event => {
-    if (this.props.onFocus) {
-      this.props.onFocus(event);
+const SwitchBase = React.forwardRef(function SwitchBase(props, ref) {
+  const {
+    autoFocus,
+    checked: checkedProp,
+    checkedIcon,
+    classes,
+    className: classNameProp,
+    defaultChecked,
+    disabled: disabledProp,
+    icon,
+    id,
+    inputProps,
+    inputRef,
+    muiFormControl,
+    name,
+    onBlur,
+    onChange,
+    onFocus,
+    readOnly,
+    required,
+    tabIndex,
+    type,
+    value,
+    ...other
+  } = props;
+  const { current: isControlled } = React.useRef(checkedProp != null);
+  const [checkedState, setCheckedState] = React.useState(Boolean(defaultChecked));
+
+  const handleFocus = event => {
+    if (onFocus) {
+      onFocus(event);
     }
 
-    const { muiFormControl } = this.props;
     if (muiFormControl && muiFormControl.onFocus) {
       muiFormControl.onFocus(event);
     }
   };
 
-  handleBlur = event => {
-    if (this.props.onBlur) {
-      this.props.onBlur(event);
+  const handleBlur = event => {
+    if (onBlur) {
+      onBlur(event);
     }
 
-    const { muiFormControl } = this.props;
     if (muiFormControl && muiFormControl.onBlur) {
       muiFormControl.onBlur(event);
     }
   };
 
-  handleInputChange = event => {
+  const handleInputChange = event => {
     const checked = event.target.checked;
 
-    if (!this.isControlled) {
-      this.setState({ checked });
+    if (!isControlled) {
+      setCheckedState(checked);
     }
 
-    if (this.props.onChange) {
-      this.props.onChange(event, checked);
+    if (onChange) {
+      onChange(event, checked);
     }
   };
 
-  render() {
-    const {
-      autoFocus,
-      checked: checkedProp,
-      checkedIcon,
-      classes,
-      className: classNameProp,
-      defaultChecked,
-      disabled: disabledProp,
-      icon,
-      id,
-      inputProps,
-      inputRef,
-      muiFormControl,
-      name,
-      onBlur,
-      onChange,
-      onFocus,
-      readOnly,
-      required,
-      tabIndex,
-      type,
-      value,
-      ...other
-    } = this.props;
+  let disabled = disabledProp;
 
-    let disabled = disabledProp;
-
-    if (muiFormControl) {
-      if (typeof disabled === 'undefined') {
-        disabled = muiFormControl.disabled;
-      }
+  if (muiFormControl) {
+    if (typeof disabled === 'undefined') {
+      disabled = muiFormControl.disabled;
     }
+  }
 
-    const checked = this.isControlled ? checkedProp : this.state.checked;
-    const hasLabelFor = type === 'checkbox' || type === 'radio';
-
-    return (
-      <IconButton
-        component="span"
-        className={clsx(
-          classes.root,
-          {
-            [classes.checked]: checked,
-            [classes.disabled]: disabled,
-          },
-          classNameProp,
-        )}
+  const checked = isControlled ? checkedProp : checkedState;
```

Comment:
Should we add a warning here too or do this in another PR?

Reply:
To warn if controlled state changes? I think this would already be called since checkedProp is passed to the checked prop, not the computed checked value. (Not sure if this is intended) ?

Reply:
Do we have a test for that? Since we removed some (bad) tests concerning controlled mode I want to make sure we are certain about the behavior.

Reply:
Yep :) https://github.com/mui-org/material-ui/blob/496b0ff106494eea8c1515906d3ed55683e309fb/packages/material-ui/src/internal/SwitchBase.test.js#L455

Reply:
Very nice. So the removed tests were not testing any behavior anyway. Probably just for line coverage.

Reply:
It was testing internals. As long as the public API works controlled vs uncontrolled isn't a big deal :)